src/monodroid/jni/android-system.cc

```diff
@@ -718,85 +763,156 @@ AndroidSystem::get_gref_gc_threshold ()
 	return static_cast<int> ((max_gref_count * 90LL) / 100LL);
 }
 
+#if defined (DEBUG) || !defined (ANDROID)
 void
-AndroidSystem::setup_environment (jstring_wrapper& name, jstring_wrapper& value)
+AndroidSystem::setup_environment (const char *name, const char *value)
 {
-	const char *k = name.get_cstr ();
-
-	if (k == nullptr || *k == '\0')
+	if (name == nullptr || *name == '\0')
 		return;
 
-	const char *v = value.get_cstr ();
-	if (v == nullptr || *v == '\0')
+	const char *v = value;
+	if (v == nullptr)
 		v = "";
 
-	if (isupper (k [0]) || k [0] == '_') {
-		if (k [0] == '_') {
-			if (strcmp (k, "__XA_DSO_IN_APK") == 0) {
-				knownEnvVars.DSOInApk = true;
-				return;
-			}
-		}
+	if (isupper (name [0]) || name [0] == '_') {
+		setenv (name, v, 1);
+		return;
+	}
+
+	add_system_property (name, v);
+}
 
-		setenv (k, v, 1);
+void
+AndroidSystem::setup_environment_from_override_file (const char *path)
+{
+	monodroid_stat_t sbuf;
+	if (utils.monodroid_stat (path, &sbuf) < 0) {
+		log_warn (LOG_DEFAULT, "Failed to stat the environment override file %s: %s", path, strerror (errno));
 		return;
 	}
 
-	if (k [0] == 'm') {
-		if (strcmp (k, "mono.aot") == 0) {
-			if (*v == '\0') {
-				knownEnvVars.MonoAOT = MonoAotMode::MONO_AOT_MODE_NONE;
-				return;
-			}
+	int fd = open (path, O_RDONLY);
+	if (fd < 0) {
+		log_warn (LOG_DEFAULT, "Failed to open the environment override file %s: %s", path, strerror (errno));
+		return;
+	}
 
-			switch (v [0]) {
-				case 'n':
-					knownEnvVars.MonoAOT = MonoAotMode::MONO_AOT_MODE_NORMAL;
-					break;
+	char    *buf = new char [sbuf.st_size];
+	ssize_t  nread = read (fd, buf, sbuf.st_size);
```

Comment:
We shouldn't assume that **read**(2) will return all data at once; `EINTR` Is A Thing™.  Thus, we may need multiple tries:

```c
ssize_t nread = 0;
do {
    int i = nread;
    nread += read (fd, buf + i, stuff.st_size - i);
} while (errno == EINTR);
```